File: library/src/main/java/me/proxer/library/api/list/IndustryListEndpoint.java
package me.proxer.library.api.list;

import lombok.Setter;
import lombok.experimental.Accessors;
import me.proxer.library.api.PagingLimitEndpoint;
import me.proxer.library.api.ProxerCall;
import me.proxer.library.entitiy.list.IndustryCore;
import me.proxer.library.enums.Country;

import javax.annotation.Nullable;
import java.util.List;

/**
 * Endpoint for retrieving a list of industries.
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public final class IndustryListEndpoint implements PagingLimitEndpoint<List<IndustryCore>> {

    private final InternalApi internalApi;

    /**
     * Sets the query to search for only from the start.
     */
    @Nullable
    @Setter
    private String searchStart;

    /**
     * Sets the query to search for.
     */
    @Nullable
    @Setter
    private String search;

    /**
     * Sets the country to filter by.
     */
    @Nullable
    @Setter
    private Country country;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter
    private Integer page;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter
    private Integer limit;

    IndustryListEndpoint(final InternalApi internalApi) {
        this.internalApi = internalApi;
    }

    @Override
    public ProxerCall<List<IndustryCore>> build() {
        return internalApi.industryList(searchStart, search, country, page, limit);

    }
}

File: library/src/main/kotlin/me/proxer/library/api/list/IndustryListEndpoint.kt
package me.proxer.library.api.list

import me.proxer.library.ProxerCall
import me.proxer.library.api.PagingLimitEndpoint
import me.proxer.library.entity.list.IndustryCore
import me.proxer.library.enums.Country

/**
 * Endpoint for retrieving a list of industries.
 *
 * @author Ruben Gees
 */
class IndustryListEndpoint internal constructor(
    private val internalApi: InternalApi
) : PagingLimitEndpoint<List<IndustryCore>> {

    private var page: Int? = null
    private var limit: Int? = null

    private var searchStart: String? = null
    private var search: String? = null
    private var country: Country? = null

    override fun page(page: Int?) = this.apply { this.page = page }
    override fun limit(limit: Int?) = this.apply { this.limit = limit }

    /**
     * Sets the query to search for only from the start.
     */
    fun searchStart(searchStart: String?) = this.apply { this.searchStart = searchStart }

    /**
     * Sets the query to search for.
     */
    fun search(search: String?) = this.apply { this.search = search }

    /**
     * Sets the country to filter by.
     */
    fun country(country: Country?) = this.apply { this.country = country }

    override fun build(): ProxerCall<List<IndustryCore>> {
        return internalApi.industryList(searchStart, search, country, page, limit)

    }
}
